### PR TITLE
[DPE-3824, DPE-3689] router exporter TLS + update mysqld exporter to v0.15

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issues
1. To rebuild the snap, we need to update the mysqld-exporter to v0.15 (as our PPA got updated)
2. We need to support TLS for mysql-router-exporter in vms

## Solutions
1. Update mysqld-exporter version that is downloaded. Update the args being passed to mysqld-exporter as breaking changes were introduced in v0.15
2. Add ability to pass router TLS cert, ca and key paths to be then passed to mysql-router-exporter

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -131,8 +131,8 @@ parts:
       - mysql-server-8.0=8.0.36-0ubuntu0.22.04.1
       - mysql-router=8.0.36-0ubuntu0.22.04.1
       - mysql-shell=8.0.36+dfsg-0ubuntu0.22.04.1~ppa4
-      - prometheus-mysqld-exporter=0.14.0-0ubuntu0.22.04.1~ppa1
-      - prometheus-mysqlrouter-exporter=4.0.5-0ubuntu0.22.04.1~ppa1
+      - prometheus-mysqld-exporter=0.15.0-0ubuntu0.22.04.1~ppa1
```
Comment (taurus-forever) [inline snap/snapcraft.yaml]: as requested, reverted to [0.14.0-0ubuntu0.22.04.1~ppa2](https://launchpad.net/~data-platform/+archive/ubuntu/mysqld-exporter/+sourcepub/15912049/+listing-archive-extra) (please notice ppa2!).

File: snap/local/start-mysqld-exporter.sh [hunk 2]
```diff
@@ -18,12 +18,13 @@ EXPORTER_PATH="/usr/bin/prometheus-mysqld-exporter"
 SOCKET="/var/run/mysqld/mysqld.sock"
 
 if [ -z "$SNAP" ]; then
-    # When not running as a snap, expect `DATA_SOURCE_NAME` to be set.
-    if [ -z "$DATA_SOURCE_NAME" ]; then
-        echo "DATA_SOURCE_NAME must be set"
+    # When not running as a snap, expect `EXPORTER_USER` and `EXPORTER_PASS` to be set.
+    if [ -z "$EXPORTER_USER" || -z "$EXPORTER_PASS" ]; then
+        echo "EXPORTER_USER and EXPORTER_PASS must be set"
```
Comment (taurus-forever) [inline snap/local/start-mysqld-exporter.sh]: Please adopt this, always post errors to STDERR:
```suggestion
        echo "ERROR: both EXPORTER_USER and EXPORTER_PASS must be set" >&2
```

File: snap/local/start-mysqld-exporter.sh [hunk 3]
```diff
@@ -18,12 +18,13 @@ EXPORTER_PATH="/usr/bin/prometheus-mysqld-exporter"
...
         exit 1
     fi
-    exec "$EXPORTER_PATH" "${EXPORTER_OPTS[@]}"
+    exec env MYSQLD_EXPORTER_PASS="${EXPORTER_PASS}" \
+        "$EXPORTER_PATH" "--mysqld.username=${EXPORTER_USER}" "--mysqld.address=unix://${SOCKET}" "${EXPORTER_OPTS[@]}"
```
Comment (taurus-forever) [inline snap/local/start-mysqld-exporter.sh]: Just layout for better readability:
```suggestion
    exec env MYSQLD_EXPORTER_PASS="${EXPORTER_PASS}" \
        "${EXPORTER_PATH}" \
        "--mysqld.username=${EXPORTER_USER}" \
        "--mysqld.address=unix://${SOCKET}" \
        "${EXPORTER_OPTS[@]}"
```

File: snap/local/start-mysqld-exporter.sh [hunk 4]
```diff
@@ -43,6 +44,6 @@ else
         --reuid snap_daemon \
         --regid snap_daemon \
         -- \
-        env DATA_SOURCE_NAME="${EXPORTER_USER}:${EXPORTER_PASS}@unix(${SOCKET})/" \
-        "$EXPORTER_PATH" "${EXPORTER_OPTS[@]}"
+        env MYSQLD_EXPORTER_PASSWORD="${EXPORTER_PASS}" \
+        "$EXPORTER_PATH" "--mysqld.username=${EXPORTER_USER}" "--mysqld.address=unix://${SOCKET}" "${EXPORTER_OPTS[@]}"
```
Comment (taurus-forever) [inline snap/local/start-mysqld-exporter.sh]:
```suggestion
        env MYSQLD_EXPORTER_PASSWORD="${EXPORTER_PASS}" \
            "${EXPORTER_PATH}" \
            "--mysqld.username=${EXPORTER_USER}" \
            "--mysqld.address=unix://${SOCKET}" \
            "${EXPORTER_OPTS[@]}"
```

File: snap/local/start-mysql-router-exporter.sh [hunk 5]
```diff
@@ -30,16 +32,32 @@ if [ -n "$SNAP" ]; then
...
+        # Execute mysqlrouter-exporter as the non-sudo user: snap-daemon.
+        exec "$SNAP"/usr/bin/setpriv \
+            --clear-groups \
+            --reuid snap_daemon \
+            --regid snap_daemon -- \
+            env MYSQLROUTER_EXPORTER_URL="${MYSQLROUTER_EXPORTER_URL}" \
+            MYSQLROUTER_EXPORTER_USER="${MYSQLROUTER_EXPORTER_USER}" \
+            MYSQLROUTER_EXPORTER_PASS="${MYSQLROUTER_EXPORTER_PASS}" \
+            "$EXPORTER_PATH" "${EXPORTER_OPTS[@]}" --skip-tls-verify
+    fi
```
Comment (taurus-forever) [inline snap/local/start-mysql-router-exporter.sh]: Not tested, but IMHO easier to read/support in the future:

```suggestion
    SETPRIV_OPT=(
        "--clear-groups"
        "--reuid snap_daemon"
        "--regid snap_daemon"
    )

    EXPORTER_ENV=(
        "MYSQLROUTER_EXPORTER_URL='${MYSQLROUTER_EXPORTER_URL}'"
        "MYSQLROUTER_EXPORTER_USER='${MYSQLROUTER_EXPORTER_USER}'"
        "MYSQLROUTER_EXPORTER_PASS='${MYSQLROUTER_EXPORTER_PASS}'"
    )
    
    if [[ -n "${MYSQLROUTER_TLS_CACERT_PATH}" && \
          -n "${MYSQLROUTER_TLS_CERT_PATH}" && \
          -n "${MYSQLROUTER_TLS_KEY_PATH}" ]]; then
        EXPORTER_ENV+=("MYSQLROUTER_TLS_CACERT_PATH='${MYSQLROUTER_TLS_CACERT_PATH}'")
        EXPORTER_ENV+=("MYSQLROUTER_TLS_CERT_PATH='${MYSQLROUTER_TLS_CERT_PATH}'")
        EXPORTER_ENV+=("MYSQLROUTER_TLS_KEY_PATH='${MYSQLROUTER_TLS_KEY_PATH}'")
    else
        EXPORTER_OPTS+=("--skip-tls-verify")
    fi
    
    # For security measures, daemons should not be run as sudo.
    # Execute mysqlrouter-exporter as the non-sudo user: snap-daemon.
    exec "${SNAP}"/usr/bin/setpriv "${SETPRIV_OPT[@]}" -- \
        env  "${EXPORTER_ENV[@]}" "${EXPORTER_PATH}" "${EXPORTER_OPTS[@]}"
```

File: snap/snapcraft.yaml (hunk 1 above)
Comment (shayancanonical) [reply snap/snapcraft.yaml]: reverted in fa4e9a8c4b064b02274bd4bfa528966a8e690469

File: snap/local/start-mysqld-exporter.sh (hunk 3 above)
Comment (shayancanonical) [reply snap/local/start-mysqld-exporter.sh]: addressed in fa4e9a8c4b064b02274bd4bfa528966a8e690469

File: snap/local/start-mysqld-exporter.sh (hunk 4 above)
Comment (shayancanonical) [reply snap/local/start-mysqld-exporter.sh]: addressed in fa4e9a8c4b064b02274bd4bfa528966a8e690469

File: snap/local/start-mysql-router-exporter.sh (hunk 5 above)
Comment (shayancanonical) [reply snap/local/start-mysql-router-exporter.sh]: in the spirit of all the readability feedback, refactored in fa4e9a8c4b064b02274bd4bfa528966a8e690469

File: snap/local/start-mysqld-exporter.sh (hunk 2 above)
Comment (shayancanonical) [reply snap/local/start-mysqld-exporter.sh]: addressed and changed all such errors to post to STDERR in fa4e9a8c4b064b02274bd4bfa528966a8e690469

**Review by carlcsaposs-canonical — APPROVED** [review]

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -131,8 +131,8 @@ parts:
       - mysql-server-8.0=8.0.36-0ubuntu0.22.04.1
       - mysql-router=8.0.36-0ubuntu0.22.04.1
       - mysql-shell=8.0.36+dfsg-0ubuntu0.22.04.1~ppa4
-      - prometheus-mysqld-exporter=0.14.0-0ubuntu0.22.04.1~ppa1
-      - prometheus-mysqlrouter-exporter=4.0.5-0ubuntu0.22.04.1~ppa1
+      - prometheus-mysqld-exporter=0.14.0-0ubuntu0.22.04.1~ppa2
+      - prometheus-mysqlrouter-exporter=5.0.1-0ubuntu0.22.04.1~ppa1
```
Comment (paulomach) [inline snap/snapcraft.yaml:135]: just to be sure, this is the version with the fix with graceful termination?

**Review by paulomach — APPROVED** [review]
LGTM

File: snap/snapcraft.yaml (hunk 6 above)
Comment (shayancanonical) [reply snap/snapcraft.yaml:135]: yes, exactly. [5.0.1](https://github.com/rluisr/mysqlrouter_exporter/blob/v5.0.1/main.go#L376) has graceful termination

**Pull request merged**